src/irmin-pack/unix/gc.ml

```diff
@@ -138,16 +138,13 @@ module Worker = struct
       in
       let buffer = Bytes.create len in
       read_exn ~off ~len buffer;
-      let entry = Mapping_file.find_nearest_leq mapping off in
-      match entry with
-      | None -> assert false
-      | Some { poff; _ } ->
-          Bytes.set buffer Hash.hash_size magic_parent;
-          (* Bytes.unsafe_to_string usage: We assume read_exn returns unique ownership of buffer
-             to this function. Then at the call to Bytes.unsafe_to_string we give up unique
-             ownership to buffer (we do not modify it thereafter) in return for ownership of the
-             resulting string, which we pass to write_exn. This usage is safe. *)
-          write_exn ~off:poff ~len (Bytes.unsafe_to_string buffer)
+      let poff = Dispatcher.poff_of_entry_exn ~off ~len mapping in
+      Bytes.set buffer Hash.hash_size magic_parent;
+      (* Bytes.unsafe_to_string usage: We assume read_exn returns unique ownership of buffer
+         to this function. Then at the call to Bytes.unsafe_to_string we give up unique
+         ownership to buffer (we do not modify it thereafter) in return for ownership of the
+         resulting string, which we pass to write_exn. This usage is safe. *)
+      write_exn ~off:poff ~len (Bytes.unsafe_to_string buffer)
```

Comment:
Good catch, that bug is scary. It completely corrupts what's written on disk.

This bug cannot happen in Tezos because in the Tezos use case the commits in the prefix are always the first objects of their mapping chunk, meaning that `entry.poff` is luckily also the `poff` of the commit.

Reply:
Could one of you describe why using `entry.poff` here is not correct? I see that we have to adjust the `entry.poff` in the dispatcher (by a calculated `shift_in_chunk`), but I don't fully follow the logic.

Reply:
So a `chunk` contains several objects and we need to adjust based on the original offset requested (`let shift_in_chunk = off_start - entry.off` from `chunk_of_off_exn`). Got it.

- In this context is `chunk` referring to a range of bytes in the prefix? And/or is it synonymous with a mapping `entry`?
- Why is the physical offset adjusted by math on virtual offsets? (looking at `chunk.poff + (off_start - entry.off)`) I'm guessing this is a result of details for how the mapping entries are created (?).

Thanks for the expanded explanation. :bow:

Reply:
> In this context is `chunk` referring to a range of bytes in the prefix? And/or is it synonymous with a mapping `entry`?

Both

> Why is the physical offset adjusted by math on virtual offsets? (looking at `chunk.poff + (off_start - entry.off)`) 

All the entries of one chunk used to be contiguous entries in a `suffix` (i.e. the global offset of the second is the global offset of the first plus the length of the first). This allows doing math from global offset to physical offset

Reply:
That was my hunch. Thanks for the clarification.

Mapping file entries can (but not always?) contain multiple objects (commits, blobs, inodes) so within a particular entry we can use virtual offset math _within_ the range of physical offsets.

Reply:
As an aside, it is confusing to my brain when using entry/chunk and entry/object interchangeably. Here is my current understanding (which I think is correct :sweat_smile:):

An `entry` maps a virtual location, `entry.off`, to a `chunk` (maybe we change this to `range` in light of upcoming chunks), `[entry.poff, entry.poff + entry.len]`, of bytes than can contain multiple `object`s.

Reply:
I'd very much be in favor of removing `entry` from the code's language since it (in my mind) implies a 1:1 relationship. It is, in a sense, but it is 1:1 from virtual offset to physical chunk/range. I was thinking of it as a 1:1 to an _object_.